src/runtime_config/default/actions.js

```diff
@@ -123,7 +123,46 @@ function checkX509CertChain(value, field) {
   }
 }
 
+function invalidateOtherOpenProposals(proposalIdToRetain) {
+  console.log(`Retaining ${proposalIdToRetain}`);
+  let proposals = ccf.kv["public:ccf.gov.proposals_info.js"];
+  const proposalsMap = ccf.kv["public:ccf.gov.proposals_info.js"];
+  proposalsMap.forEach((v, k) => {
+    let proposalId = ccf.bufToStr(k);
+    console.log(`Looking at ${proposalId}`);
+    if (proposalId !== proposalIdToRetain) {
+      let info = ccf.bufToJsonCompatible(v);
+      if (info.state === "Open") {
+        console.log(`Setting state of ${proposalId} to invalidated`);
+        info.state = "Invalidated";
+        proposalsMap.set(k, ccf.jsonCompatibleToBuf(info));
+      } else {
+        console.log(`State is ${info.state} - ignoring`);
+      }
+    } else {
+      console.log("Match, making no changes");
+    }
+  });
+}
+
 const actions = new Map([
+  [
+    "set_constitution",
+    new Action(
+      function (args) {
+        checkType(args.constitution, "string");
+      },
+      function (args, proposalId) {
+        ccf.kv["public:ccf.gov.constitution"].set(
+          getSingletonKvKey(),
+          ccf.strToBuf(JSON.stringify(args.constitution))
```

Comment:
```suggestion
          ccf.jsonCompatibleToBuf(args.constitution)
```

Reply:
Why? constitution is a string.

Reply:
Yeah, but all ServiceMap's are JSON values.